add docs for zed language (#3708)
brimdata/zed: docs/zq/language.md

```
a Zed shaper could catch the bad value (e.g., as a default case in a
switch topology) and propagate it as an error using the Zed expression:
```
yield error({message:"unrecognized input",input:this})
```
then such errors could be detected and searched for downstream with the
[is_error function](functions/is_error.md).
For example,
```
is_error(this)
```
on the wrapped error from above produces
```
error({message:"unrecognized input",input:{kind:"bad", stuff:{foo:1,bar:2}}})
```
There is no need to create special tables in a complex warehouse-style ETL
to land such errors as they can simply land next to the output values themselves.

And when transformations cascade one into the next as different stages of
an ETL pipeline, errors can be wrapped one by one forming a "stack trace"
or lineage of where the error started and what stages it traversed before
landing at the final output stage.

Errors will unfortunately and inevitably occur even in production,
but having a first-class data type to manage them all while allowing them to
peacefully co-exist with valid production data is a novel and
useful approach that Zed enables.

#### 5.3.1 Missing and Quiet

> Zed's heterogenous data model allows for queries

Reviewer: ```suggestion
Zed's heterogeneous data model allows for queries
```

Thus, a reference to a field, e.g., `this.x` may be valid for some values
that include a field called `x` but not valid for those that do not.

What is the value of `x` when the field `x` does not exist?

A similar question faced SQL when it was adapted in various different forms
to operate on semi-structured data like JSON or XML.  SQL already had the `NULL` value
so perhaps a reference to a missing value could simply be `NULL`.

But JSON also has `null`, so a reference to `x` in the JSON value
```
{"x":null}
```
and a reference to `x` in the JSON value
```
{}
```
would have the same value of `NULL`.  Furthermore, an expression like `x==NULL`
could not differentiate between these two cases.

To solve this problem, the `MISSING` value was proposed to represent the value that
results from accessing a field that is not present.  Thus, `x==NULL` and
`x==MISSING` could disambiguate the two cases above.

Zed, instead, recognizes that the SQL value is `MISSING` is a paradox:
I'm here but I'm not.  

In reality, a MISSING value is not a value.  It's an error condition
that resulted from trying to reference something that didn't exist.

> So why should pretend that this is a bona fide value?  SQL adopted this

Reviewer: ```suggestion
So why should we pretend that this is a bona fide value?  SQL adopted this
```

`error("missing")`.

### 6.6 Indexing

The index operation can be applied to various data types and has the form:
```
<value> [ <index> ]
```
If the `<value>` expression is a record, then the `<index>` operand
must be coercible to a string and the result of the record's field
of that name.

If the `<value>` expression is an array, then the `<index>` operand
must be coercible to an integer and the result is the
value in the array of that index.

If the `<value>` expression is a set, then the `<index>` operand
must be coercible to an integer and the result is the
value in the set of that index ordered by total order of Zed values.

If the `<value>` expression is a map, then the `<index>` operand
is presumed to be a key and the corresponding value for that key is
the result of the operation.  If no such key exists in the map, then
`error("missing")` results.

If the `<value>` expression is a string, then the `<index>` operand
must be coercible to an integer and the result is an integer representing
the unicode code point at that offset in the string.

If the `<value>` expression is type bytes, then the `<index>` operand

> must be coercible to an integer and the result is a unsigned 8-bit integer

Reviewer: ```suggestion
must be coercible to an integer and the result is an unsigned 8-bit integer
```

representing the byte value at that offset in the bytes sequence.

### 6.7 Slices

The slice operation can be applied to various data types and has the form:
```
<value> [ <from> : <to> ]
```
The `<from>` and `<to>` terms must be expressions that are coercible
to integer and represent a range of index values to form a subset of elements
from the `<value>` term provided.  The range begins at the `<from>` position
and ends one before the `<to>` position.  A negative
value of `<from>` or `<to>` represents a position relative to the
end of the value being sliced.

If the `<value>` expression is an array, then the result is an array of
elements comprising the indicated range.

> If the `<value>` expression is an set, then the result is a set of

Reviewer: ```suggestion
If the `<value>` expression is a set, then the result is a set of
```

elements comprising the indicated range ordered by total order of Zed values.

> If the `<value>` expression is an string, then the result is a substring

Reviewer: ```suggestion
If the `<value>` expression is a string, then the result is a substring
```

consisting of unicode code points comprising the given range.

If the `<value>` expression is type bytes, then the result is a bytes sequence
consisting of bytes comprising the given range.

### 6.8 Conditional

A conditional expression has the form
```
<boolean> ? <expr> : <expr>`
```

> The `<boolean>` expression is evaluated and must be result it type `bool`.

Reviewer: ```suggestion
The `<boolean>` expression is evaluated and the result must be of type `bool`.
```

Author: must have a result of type `bool`.

the result.  Otherwise, the second `<expr>` expression is evaluated.

Note that if the expression has side effects,
as with [aggregation calls](#aggregation-calls), only the selected expression
will be evaluated.

For example,
```mdtest-command
echo '{s:"foo",v:1}{s:"bar",v:2}' | zq -z 'yield (s=="foo") ? v : -v' -
```
produces
```mdtest-output
1
-2
```

### 6.9 Function Calls

Functions perform stateless transformations of their input value to their
return value and utilize call-by value semantics with positional and unnamed
arguments.  Some functions take a variable number of arguments.

> The only available functions are built-in but user-defined functions and
> library package management will be added to the Zed language soon.

For example,
```mdtest-command
zq -z 'yield pow(2,3), lower("ABC")+upper("def"), typeof(1)'
```
produces
```mdtest-output
8.
"abcDEF"
<int64>
```

### 6.10 Aggregate Function Calls

[Aggregate functions](reference.md#aggregates) may be called within an expression.
Unlike the aggregation context provided by a summarizing group-by, such calls
in expression context yield an output value for each input value.

> Note that because aggregate functions carry state which is typically dependent
> dependent on the order of input values, their use can prevent the runtime

Reviewer: ```suggestion
Note that because aggregate functions carry state which is typically
dependent on the order of input values, their use can prevent the runtime
```

The second form is an implied field reference `<ref>`, which is shorthand
for `<ref>:<ref>`.  The third form is the `...` spread operator which expects
a record value as the result of `<expr>` and inserts all of the fields from
the resulting record.
If a spread expression results in a non-record type (e.g., errors), then that
part of record is simply elided.

The fields of a record expression are evaluated left to right and when
field names collide the rightmost instance of the name determines that
field's value.

For example,
```mdtest-command
echo '{x:1,y:2,r:{a:1,b:2}}' | zq -z 'yield {a:0},{x}, {...r}, {a:0,...r,b:3}' -
```
produces
```mdtest-output
{a:0}
{x:1}
{a:1,b:2}
{a:1,b:3}
```

#### 6.11.3 Array Expressions

Array literals have the form
```
[ <expr>, <expr>, ... ]
```
When the expressions result in values of non-uniform type, then the implied
type of the array is an array of type union of the types that appear.

For example,
```mdtest-command
zq -z 'yield [1,2,3],["hello","world"]'
```
produces
```mdtest-output
[1,2,3]
["hello","world"]
```

#### 6.11.4 Set Expressions

Set literals have the form
```
|[ <expr>, <expr>, ... ]|
```

> When the expressions results in values of non-uniform type, then the implied

Reviewer: ```suggestion
When the expressions result in values of non-uniform type, then the implied
```

type of the set is a set of type union of the types that appear.

Set values are always organized in their "natural order" independent of the order
they appear in the set literal.

For example,
```mdtest-command
zq -z 'yield |[3,1,2]|,|["hello","world","hello"]|'
```
produces
```mdtest-output
|[1,2,3]|
|["hello","world"]|
```

#### 6.11.5 Map Expressions

Map literals have the form
```
|{ <expr>:<expr>, <expr>:<expr>, ... }|
```
where the first expression of each colon-separated entry is the key value
and the second expression is the value.
When the key and/or value expressions results in values of non-uniform type,
then the implied type of the map has a key type and/or value type that is
a union of the types that appear in each respective category.

For example,
```mdtest-command
zq -z 'yield |{"foo":1,"bar"+"baz":2+3}|'
```
produces
```mdtest-output
|{"foo":1,"barbaz":5}|
```

#### 6.11.6 Union Values

A union value can be created with a cast, e.g.,
```mdtest-command
echo '1 "foo"' | zq -z 'yield cast(this,<(int64,string)>)' -
```
produces
```mdtest-output
1((int64,string))
"foo"((int64,string))
```

> The value underlying a union-tagged value is accesse with the

Reviewer: ```suggestion
The value underlying a union-tagged value is accessed with the
```

[under function](functions/under.md):
```mdtest-command
echo '1((int64,string))' | zq -z 'yield under(this)' -
```
produces
```mdtest-output
1
```

### 6.12 Constants

Constants may be declared and bound to an identifier with a `const` statement,
which has the form
```
const <id> = <expr>
```
This statement evaluates the expressions `<expr>` before any input is processed
to determine the constant value assigned to the identifier `<id>`.
This expression may not refer to `this` or any implied fields of `this`.

Constants must appear at the beginning
of a query or the beginning of a [parenthesized scope](#scopes).

```mdtest-command
echo '{diameter:1}{diameter:5}' | zq -z 'const PI = 3.14159 circumference:=2*PI*diameter' -
```
produces
```mdtest-output
{diameter:1,circumference:6.28318}
{diameter:5,circumference:31.4159}
```

### 6.13 Type Definitions

Named types may be declared and bound to an identifier with a `type` statement,
which has the form
```
type <name> = <type>
```
This statement evaluates the type `<type>` before any input is processed

> to determine the a type value assigned to the identifier `<id>`.

Reviewer: ```suggestion
to determine the type value assigned to the identifier `<id>`.
```

> In this case, grep will still support substring match but keyword search
> will match segmented words from string fields so that they can be efficiently
> queried in search indexes.

##### 7.2.1.5 Non-String Literal Search Term

Search terms representing non-string Zed values search for both an exact
match for the given value as well as a string search for the term exactly
as it appears as typed.  Such values include:
* integers,
* floating point numbers,
* time values,
* durations,
* IPs,
* networks,
* bytes values, and
* type values.

A search for a Zed value `<value>` represented as the string `<string>` is
equivalent to
```
<value> in this or grep(<string>, this)
```
For example,
```
search 123 and 10.0.0.1
```
which can be abbreviated
```
123 10.0.0.1
```
is equivalent to
```
where (123 in this or grep("123", this)) and (10.0.0.1 in this or grep("10.0.0.1", this))
```

Complex values are not supported as search terms but may be queried with
the "in" operator, e.g.,
```
{s:"foo"} in this
```

##### 7.2.1.6 Predicate Search Term

Any Boolean-valued [function](reference.md) like `is()`, `has()`,
`grep()` etc. and any [comparison expression](#comparisons)

> may be used a search term and mixed into a search expression.

Reviewer: ```suggestion
may be used as a search term and mixed into a search expression.
```

to each subsequence of values generated from an outer sequence of values.
The inner query may be _any Zed query_ and may refer to values from
the outer sequence.

Lateral queries are created using the scoped form of the
[over operator](operators/over.md#operator) and may be nested to arbitrary depth.

For example,
```mdtest-command
echo '{s:"foo",a:[1,2]} {s:"bar",a:[3]}' | zq -z 'over a with name=s => (yield {name,elem:this})' -
```
produces
```mdtest-output
{name:"foo",elem:1}
{name:"foo",elem:2}
{name:"bar",elem:3}
```
You can also import a parent-scope field reference into the inner scope by
simply referring to its name without assignment, e.g.,
```mdtest-command
echo '{s:"foo",a:[1,2]} {s:"bar",a:[3]}' | zq -z 'over a with s => (yield {s,elem:this})' -
```
produces
```mdtest-output
{s:"foo",elem:1}
{s:"foo",elem:2}
{s:"bar",elem:3}
```

### 8.1 Lateral Scope

A lateral scope has the form `=> ( <query> )` and currently appears
only the context of an [over operator](operators/over.md#operator):
```
over ... with <elem> [, <elem> ...] => ( <query> )
```
where `<elem>` has either an assignment form
```
<var>=<expr>
```
or a field reference form
```
<field>
```

> For each input value to the outer scope, the assignment form create a binding

Reviewer: ```suggestion
For each input value to the outer scope, the assignment form creates a binding
```

```
drops the `uid` field after shaping:
```mdtest-output
{
    kind: "dns",
    client: {
        addr: 10.47.1.100,
        port: 41772 (port=uint16)
    } (=socket),
    server: {
        addr: 10.0.0.100,
        port: 53
    } (socket),
    vlan: null (uint16)
}
```
## 10. Type Fusion

Type fusion is another important building block of data shaping.
Here, types are operated upon by fusing them together, where the
result is single fused type.
Some systems call a related process "schema inference" where a set
of values, typically JSON, is analyzed to determine a relational schema
that all the data will fit into.  However, this is just a special case of
type fusion as fusion is fine-grained and based on Zed's type system rather
than having the narrower goal of computing a schema for representations
like relational tables, Parquet, Avro, etc.

Type fusion utilized two key techniques.

The first technique is to simply combine types with a type union.
For example, an `int64` and a `string` can be merged into a common
type of union `(int64,string)`, e.g., the value sequence `1 "foo"`
can be fused into the single-type sequence:
```
1((int64,string))
"foo"((int64,string))
```

> The second technique is to merge fields of records, analagous to a spread

Reviewer: ```suggestion
The second technique is to merge fields of records, analogous to a spread
```

expression.  Here, the value sequence `{a:1}{b:"foo"}` may be
fused into the single-type sequence:
```
{a:1,b:null(string)}
{a:null(int64),b:"foo"}
```

Of course, these two techniques can be powerfully combined,
e.g., where the value sequence `{a:1}{a:"foo",b:2}` may be
fused into the single-type sequence:
```
{a:1((int64,string)),b:null(int64)}
{a:"foo"((int64,string)),b:2}
```

To perform fusion, Zed currently includes two key mechanisms
(though this is an active area of development):
* the [fuse operator](operators/fuse.md#operator), and
* the [fuse aggregate function](aggregates/fuse.md).

### 10.1 Fuse Operator

The _fuse operator_ reads all of its input, computes a fused type using

> the techniuqes above, and outputs the result, e.g.,

Reviewer: ```suggestion
the techniques above, and outputs the result, e.g.,
```

* [crop](functions/crop.md) - remove fields from a value that are missing in a specified type
* [fill](functions/fill.md) - add null values for missing fields
* [order](functions/order.md) - reorder record fields
* [shape](functions/shape.md) - apply cast, fill, and order

They all have the same signature, taking two parameters: the value to be
transformed and a type value for the target type.

> Another type of transformation that's needed for shaping is renaming fields,
> which is supported by the [rename operator](operators/rename.md#operator).
> Also, the [yield operator](operators/yield.md#operator)
> is handy for simply emitting new, arbitrary record literals based on
> input values and mixing in these shaping functions in an embedded record literal.
> The [fuse aggregate function](aggregates/fuse.md) is useful for fusing
> values into a common schema within a group-by schema though a type is returned
> rather than values.

In the examples below, we will use the following named type `connection`
that is stored in a file [connection.zed](connection.zed)
and is included in the example Zed queries with `-I` option of `zq`:

> ```

Reviewer 2: For example files like this, we should use the content here in the Markdown rather than duplicate it in a separate file.

To do that, change `mdtest-command dir=docs/zq` to just `mdtest-command` everywhere and then add `mdtest-input <filename>` to the fenced code blocks for the example files, like this:
~~~suggestion
```mdtest-input connection.zed
~~~

Author: cool!

Author: done

&mdash; e.g., often implementing a Zed program differently than
the flow implied by the pipeline yet reaching the same result &mdash;
much as a modern SQL engine optimizes a declarative SQL query.

Zed is also intended to provide a seamless transition from a simple search experience
(e.g., typed into a search bar or as the query argument of the  `zq` command-line
tool) to more a complex analytics experience composed of complex joins and aggregations
where the Zed language source text would typically be authored in a editor and
managed under source-code control.

Like an email or Web search, a simple keyword search is just the word itself,
e.g.,
```
sample.com
```
is a search for the string "sample.com" and
```
sample.com urgent
```
is a search for values with the both strings "sample.com" and "urgent" present.

Unlike typical log search systems, the Zed language operators are uniform:
you can specify an operator including keyword search terms, Boolean predicates,
etc. using the same syntax at any point in the pipeline as
[described below](#7-search-expressions)

For example,
the predicate `message_length > 100` can simply be tacked onto the keyword search
from above, e.g.,
```

> sample.com urgent message_length > 100

Reviewer: ```suggestion
example.com urgent message_length > 100
```
or,
```suggestion
urgent sample.com message_length > 100
```
or,
```suggestion
'sample.com' urgent message_length > 100
```
(`sample` is reserved for the zed operator - renaming to `example.com` might make most sense?)

```
finds all values containing the string "sample.com" and "urgent" somewhere in them
provided further that the field `message_length` is a numeric value greater than 100.
A related query that performs an aggregation could be more formally
written as follows:
```
search "sample.com" AND "urgent"
| where message_length > 100
| summarize kinds:=union(type) by net:=network_of(srcip)
```
which computes an aggregation table of different message types (e.g.,
from a hypothetical field called `type`) into a new, aggregated field
called `kinds` and grouped by by the network of all the source IP address
in the input
(e.g.,from a hypothetical field called `srcip`) as a derived field called `net`.

The short-hand query from above might be typed into a search box while the
latter query might be composed in a query editor or in Zed source files

> maintained in Github.  Both forms are valid Zed queries.

Reviewer: ```suggestion
maintained using git ("gitops for data").  Both forms are valid Zed queries.
```

Author: I'm not sure Noah would go for this :-)

Reviewer 2: I think either "in GitHub" or "in Git" is fine, though my heart lies with the more generic "in a version control system".

e.g.,
```
sort s
```
is shorthand for `sort this.s`

### 2.5 Field Assignments

A typical operation in records involves
adding or changing the fields of a record using the [put operator](operators/put.md#operator)
or extracting a subset of fields using the [cut operator](operators/cut.md#operator).
Also, when aggregating data using group-by keys, the group-by assignments
create new named record fields.

In all of these cases, the Zed language uses the token `:=` to denote
field assignment.  For example,
```
put x:=y+1
```
or
```
summarize salary:=sum(income) by address:=lower(address)
```
This style of "assignment" to a record value is distinguished from the `=`
token which binds a locally scoped name to a value that can be referenced
in later expressions.

### 2.6 Implied Operators

When Zed is run in an application like [Brim](https://github.com/brimdata/brim),
queries are often composed interactively in a "search bar" experience.
The language design here attempts to support both this "lean forward" pattern of usage
along with a "coding style" of query writing where the queries might be large
and complex, e.g., to perform transformations in a data pipeline, where

> the Zed queries are stored under source-code control perhaps in Github or

Reviewer 2: ```suggestion
the Zed queries are stored under source-code control perhaps in GitHub or
```

```
count() by id
```
Likewise, a very common lean-forward use pattern is "searching" so by default,
expressions are interpreted as keyword searches, e.g.,
```
search foo bar or x > 100
```
is abbreviated
```
foo bar or x > 100
```
Furthermore, if an operator-free expression is not valid syntax for
a search expression but is a valid [Zed expression](#expressions),
then the abbreviation is treated as having an implied `yield` operator, e.g.,
```
{s:lower(s)}
```
is shorthand for
```
yield {s:lower(s)}
```
When operator names are omitted, `search` has precedence over `yield`, so
```
foo
```
is interpreted as a search for the string "foo" rather than a yield of
the implied record field named `foo`.

Another common query pattern involves adding or mutating fields of records
where the input is presumed to be a sequence of records.
The [put operator](operators/put.md#operator) provides this mechanism and the `put`
keyword is implied by the mutator syntax `:=`, which is used in Zed when an
input record field is modified, as compared to `=` which is used in constant
and variable assignments.  For example, the operation
```
put y:=2*x+1
```
can be expressed simply as
```
y:=2*x+1
```

> When composing long-form queries that are shared via Brim or managed in Github,

Reviewer 2: ```suggestion
When composing long-form queries that are shared via Brim or managed in GitHub,
```

In summary, if no operator name is given, the implied operator is determined
from the operator-less source text, in the order given, as follows:
* If the text can be interpreted as a search expression, then the operator is `search`.
* If the text can be interpreted as a boolean expression, then the operator is `where`.
* If the text can be interpreted as one or more field assignments, then the operator is `put`.
* If the text can be interpreted as an aggregation, then the operator is `summarize`.
* If the text can be interpreted as an expression, then the operator is `yield`.
* Otherwise, the text causes a compile-time error.

When in doubt, you can always check what the compiler is doing under the hood
by running `zq` with the `-C` flag to print the parsed query in "canonical form", e.g.,
```mdtest-command
zq -C foo
zq -C 'is(<foo>)'
zq -C 'count()'
zq -C '{a:x+1,b:y-1}'
zq -C 'a:=x+1,b:=y-1'
```
produces
```mdtest-output
search foo
where is(<foo>)
summarize
    count()
yield {a:x+1,b:y-1}
put a:=x+1,b:=y-1
```

## 3. Const Statements

Constants may be defined and assigned to a symbolic name with the syntax
```

> const <id> = <expr>

Reviewer 2: To avoid confusion with Zed type values, I think we should adopt a convention other than angle brackets for syntax examples like this. Caps or italics, maybe?

Author: I agree but this is used everywhere.  I created #3270

```

> where `<id>` is an identifier and `<expr>` or a constant [expression](#6-expressions)

Reviewer 2: ```suggestion
where `<id>` is an identifier and `<expr>` is a constant [expression](#6-expressions)
```

that must evaluate to a constant and at compile time and not reference any
runtime state like `this`, e.g.,
```mdtest-command
echo '{r:5}{r:10}' | zq -z "const PI=3.14159 2*PI*r" -
```
produces
```mdtest-output
31.4159
62.8318
```

One or more const statements may appear only at the beginning of a scope
(i.e., the main scope at the start of a Zed program or a [lateral scope](#81-lateral-scope)
defined by an [over operator](operators/over.md#operator))
and binds the identifier to the value in the scope in which it appears in addition
to any contained scopes.

A const statement cannot redefine an identifier that was previously defined in the same
scope but can override identifiers defined in ancestor scopes.

Const statements may appear intermixed with type statements.

## 4. Type Statements

Named types may be created with the syntax
```
type <id> = <type>
```
where `<id>` is an identifier and `<type>` or a [Zed type](#first-class-types).

> This create a new type with the given name in the Zed test system, e.g.,

Reviewer 2: ```suggestion
This create a new type with the given name in the Zed type system, e.g.,
```

any Zed type may be used as a value.

The primitive types are listed in the
[data model specification](../formats/zed.md#1-primitive-types)
and have the same syntax in the Zed language.  
Complex types also follow the ZSON syntax.  Note that the type of a type value
is simply `type`.

As in ZSON, _when types are used as values_, e.g., in a Zed expression,
they must be referenced within angle brackets.  That is, the integer type
`int64` is expressed as a type value using the syntax `<int64>`.

Complex types in the Zed language follow the ZSON syntax as well.  
Here are a few examples:
```
* a simple record type - `{x:int64,y:int64}
* an array of integers - `[int64]`
* a set of strings - `|[string]|`
* a map of strings keys to integer values - `{[string,int64]}`
* a union of string and integer  - `(string,int64)`
```
Complex types may be composed, as in `[({s:string},{x:int64})]` which is
an array of type union of two types of records.

The [typeof function](functions/typeof.md) returns a value's type as
a value, e.g., `typeof(1)` is `<int64>` and `typeof(<int64>)` is `<type>`.

First-class types are quite powerful because types can

> serve as group-by keys or used in "data shaping" logic.

Reviewer 2: ```suggestion
serve as group-by keys or be used in "data shaping" logic.
```

```
produces
```mdtest-output
{typeof:<int64>,count:2(uint64)}
{typeof:<string>,count:1(uint64)}
{typeof:<ip>,count:1(uint64)}
{typeof:<type>,count:1(uint64)}
```
When running such a query over complex, semi-structured data, the results can
be quite illuminating and can inform the design of "data shaping" Zed queries
to transform raw, messy data into clean data for downstream tooling.

Note the somewhat subtle difference between a record value with a field `t` of
type `type` whose value is type `string`
```
{t:<string>}
```
and a record type used as a value
```
<{t:string}>
```

### 5.2 Named Types

As in any modern programming language, types can be named and the type names
persist into the data model and thus into the serialized input and output.

Types are defined with a type statement of the form
```
type <name> = <type>
```
where `<name>` is typically an identifier (though a type name can have any
characters in it so `<name>` can also be a quoted string), and `<type>` is
any Zed type.  For example,
```
type port=uint16
```
defines a type named "port" for type `uint16`.

Type statements (along with const statements) must appear at the beginning

> of a query or the beginning of a [parenthesized scope](#scopes).

Reviewer 2: This section seems to repeat material in section 4.

Author: Indeed, will rework.

name referred to this way does not exist, then the type value reference
results in `error("missing")`.  For example,
```mdtest-command
echo '1(=foo) 2(=bar) 3(=foo)' | zq -z 'typeof(this)==<foo>' -
```
results in
```mdtest-output
1(=foo)
3(=foo)
```
and
```mdtest-command
echo '1(=foo)' | zq -z 'yield <foo>' -
```
results in
```mdtest-output
<foo=int64>
```
but
```mdtest-command
zq -z 'yield <foo>'
```
gives
```mdtest-output
error("missing")
```
Each instance of a named type definition overrides any earlier definition.
In this way, types are local in scope.

Each value that references a named type retains its local definition of the
named type retaining the proper type binding while accommodating changes in a
particular named type.  For example,
```mdtest-command
echo '1(=foo) 2(=bar) "hello"(=foo) 3(=foo)' | zq -z 'count() by typeof(this) | sort this' -
```
results in
```mdtest-output
{typeof:<bar=int64>,count:1(uint64)}
{typeof:<foo=int64>,count:2(uint64)}
{typeof:<foo=string>,count:1(uint64)}
```
Here, the two versions of type "foo" are retained in the group-by results.

In general, it is bad practice to define multiple versions a single named type

> though the Zed system and Zed data model accommodates such dynamic bindings.

Reviewer 2: ```suggestion
though the Zed system and Zed data model accommodate such dynamic bindings.
```

```
from anywhere | is(<employee>) | cut last,salary | sort salary | head 5
```
The power of Zed is that you can interpret data on the fly as belonging to
a certain schemas, in this case "employee", and those records can be intermixed
with other relevant data.  There is no need to create a table called "employee"
and put the data into the table before that data can be queried as an "employee".
And if the schema or type name for "employee" changes, queries still continue
to work.

### 5.3 First-class Errors

As with types, errors in Zed are first-class: any value can be transformed
into an error by wrapping it in the Zed [error type](../formats/zed.md#27-error).

In general, expressions and functions that result in errors simply return
a value of type error as a result.  This encourages a powerful flow-style
of error handling where errors simply propagate from one operation to the
next and land in the output alongside non-error values to provide a very helpful
context and rich information for tracking down the source of errors.  There is
no need to check for error conditions everywhere or look through auxiliary
logs to find out what happened.

For example,

> the following values can be transformed to errors as follows:

Reviewer 2: ```suggestion
input values can be transformed to errors as follows:
```

```mdtest-command
echo '0 "foo" 10.0.0.1' | zq -z 'error(this)' -
```
produces
```mdtest-output
error(0)
error("foo")
error(10.0.0.1)
```
More practically, errors from the runtime show up as error values.
For example,
```mdtest-command
echo 0 | zq -z '1/this' -
```
produces
```mdtest-output
error("divide by zero")
```
And since errors are first-class and just values, they have a type.
In particular, they are a complex type where the error value's type is the
complex type error containing the type of the value.  For example,
```mdtest-command
echo 0 | zq -z 'typeof(1/this)' -
```
produces
```mdtest-output
<error(string)>
```
First-class errors are particularly useful for creating structured errors.
When a Zed query encounters a problematic condition,
instead of silently dropping the problematic error
and logging an error obscurely into some hard-to-find system log as so many
ETL pipelines do, the Zed logic can
preferably wrap the offending value as an error and propagate it to its output.

> For example, suppose a bad value shows up...

Reviewer 2: ```suggestion
For example, suppose a bad value shows up:
```

```
{kind:"bad", stuff:{foo:1,bar:2}}
```

> a Zed shaper could catch the bad value (e.g., as a default case in a

Reviewer 2: ```suggestion
A Zed shaper could catch the bad value (e.g., as a default case in a
```

switch topology) and propagate it as an error using the Zed expression:
```
yield error({message:"unrecognized input",input:this})
```

> then such errors could be detected and searched for downstream with the

Reviewer 2: ```suggestion
Then such errors could be detected and searched for downstream with the
```

[is_error function](functions/is_error.md).
For example,
```
is_error(this)
```
on the wrapped error from above produces
```
error({message:"unrecognized input",input:{kind:"bad", stuff:{foo:1,bar:2}}})
```
There is no need to create special tables in a complex warehouse-style ETL
to land such errors as they can simply land next to the output values themselves.

And when transformations cascade one into the next as different stages of
an ETL pipeline, errors can be wrapped one by one forming a "stack trace"
or lineage of where the error started and what stages it traversed before
landing at the final output stage.

Errors will unfortunately and inevitably occur even in production,
but having a first-class data type to manage them all while allowing them to

> peacefully co-exist with valid production data is a novel and

Reviewer 2: ```suggestion
peacefully coexist with valid production data is a novel and
```

```
{"x":null}
```
and a reference to `x` in the JSON value
```
{}
```
would have the same value of `NULL`.  Furthermore, an expression like `x==NULL`
could not differentiate between these two cases.

To solve this problem, the `MISSING` value was proposed to represent the value that
results from accessing a field that is not present.  Thus, `x==NULL` and
`x==MISSING` could disambiguate the two cases above.

Zed, instead, recognizes that the SQL value is `MISSING` is a paradox:
I'm here but I'm not.  

In reality, a MISSING value is not a value.  It's an error condition
that resulted from trying to reference something that didn't exist.

So why should we pretend that this is a bona fide value?  SQL adopted this
approach because it lacks first-class errors.

But Zed has first-class errors so
a reference to something that does not exist is an error of type
`error<string>` whose value is `error("missing")`.  For example,
```mdtest-command
echo "{x:1} {y:2}" | zq -z 'yield x' -
```
produces
```mdtest-output
1
error("missing")
```
Sometimes you want missing errors to show up and sometimes you don't.
The [quiet function](functions/quiet.md) transforms missing errors into

> "quiet errors".  A quiet error is value `error("quiet")` and is ignored

Reviewer 2: ```suggestion
"quiet errors".  A quiet error is the value `error("quiet")` and is ignored
```

A comparison expression is any valid Zed expression compared to any other
valid Zed expression using a comparison operator.

When the operands are coercible to like types, the result is the truth value
of the comparison.  Otherwise, the result is `false`.

If either operand to a comparison
is `error("missing")`, then the result is `error("missing")`.

For example,
```mdtest-command
zq -z 'yield 1 > 2, 1 < 2, "b" > "a", 1 > "a", 1 > x'

```
produces
```mdtest-output
false
true
true
false
error("missing")
```

### 6.3 Containment

The `in` operator has the form
```
<item-expr> in <container-expr>
```
and is true if the `<item-expr>` expression results in a value that
appears somewhere in the `<container-expr>` as an exact match of the item.
The right-hand side value can be any Zed value and complex values are
recursively traversed to determine if the item is present anywhere within them.

For example,
```mdtest-command
echo '{a:[1,2]}{b:{c:3}}{d:{e:1}}' | zq -z '1 in this' -
```
produces
```mdtest-output
{a:[1,2]}
{d:{e:1}}
```

### 6.4 Logic

The keywords `and`, `or`, and `not` perform logic on operands of type `bool`.

> The binary operands of `and` and `or` operate on Boolean values and result in

Reviewer 2: ```suggestion
The binary operators `and` and `or` operate on Boolean values and result in
```

> an error value if either operand is not a boolean.  Likewise, `not` operates

Reviewer 2: ```suggestion
an error value if either operand is not a Boolean.  Likewise, `not` operates
```

on its unary operand and results in an error if its operand is not type `bool`.
Unlike many other languages, non-Boolean values are not automatically converted to
Boolean type using "truthiness" heuristics.

### 6.5 Field Dereference

Records fields are dereferenced with the dot operator `.` as is customary
in other languages and have the form
```
<value> . <id>
```
where `<id>` is an identifier representing the field name referenced.
If a field name is not representable as an identifier, then [indexing](#indexing)

> may be used with a quoted to string to represent any valid field name.

Reviewer 2: ```suggestion
may be used with a quoted string to represent any valid field name.
```

```
produces
```mdtest-output
1
```
Union values are powerful because they provide a mechanism to precisely
describe the type of any nested, semi-structured value composed of elements
of different types.  For example, the type of the value `[1,"foo"]` in Javascript
is simply a generic Javascript "object".  But in Zed, the type of this
value is an array of union of string and integer, e.g.,
```mdtest-command
echo '[1,"foo"]' | zq -z 'typeof(this)' -
```
produces
```mdtest-output
<[(int64,string)]>
```

### 6.12 Constants

Constants may be declared and bound to an identifier with a `const` statement,
which has the form
```
const <id> = <expr>
```
This statement evaluates the expressions `<expr>` before any input is processed
to determine the constant value assigned to the identifier `<id>`.
This expression may not refer to `this` or any implied fields of `this`.

Constants must appear at the beginning
of a query or the beginning of a [parenthesized scope](#scopes).

```mdtest-command
echo '{diameter:1}{diameter:5}' | zq -z 'const PI = 3.14159 circumference:=2*PI*diameter' -
```
produces
```mdtest-output
{diameter:1,circumference:6.28318}
{diameter:5,circumference:31.4159}
```

> ### 6.13 Type Definitions

Reviewer 2: These sections largely duplicate sections 3 and 4, but maybe that's OK.

(as well as the [grep function](functions/grep.md)) using "patterns",
where a pattern is a regular expression, glob, or simple string.

#### 7.1.1 Regular Expressions

A regular expression is specified in the familiar slash syntax where the
expression begins with a `/` character and ends with a terminating `/` character.
The string between the slashed (exclusive of those characters) is the
regular expression.

The format of Zed regular expressions follows the syntax of the
[RE2 regular expression library](https://github.com/google/re2)
and is documented in the
[RE2 Wiki](https://github.com/google/re2/wiki/Syntax).

Regular expressions may be used freely in search expressions, e.g.,
```mdtest-command
echo '"foo" {s:"bar"} {s:"baz"} {foo:1}' | zq -z '/(foo|bar)/' -
```
produces
```mdtest-output
"foo"
{s:"bar"}
{foo:1}
```
Regular expressions may also appear in the `grep` function:
```mdtest-command
echo '"foo" {s:"bar"} {s:"baz"} {foo:1}' | zq -z 'yield grep(/ba.*/, s)' -
```
produces
```mdtest-output
false
true
true
false
```

#### 7.1.2 Globs

Globs provide a convenient short-hand for regular expressions and follow
the familiar pattern of "file globbing" supported by Unix shells.

> Zed globs are a simple, special case utilize only the `*` wildcard.

Reviewer 2: ```suggestion
Zed globs are a simple, special case utilizing only the `*` wildcard.
```

any valid string escape sequence
(along with escapes for `*`, `=`, `+`, `-`), and the unescaped characters:
```
_ . : / % # @ ~
```
A glob must begin with one of these characters or `*` then may be
followed by any of these characters, `*`, or digits `0` through `9`.

> Note that these rules do not allow for a leading digit.

For example, a prefix match is easily accomplished via `prefix*`, e.g.,
```mdtest-command
echo '"foo" {s:"bar"} {s:"baz"} {foo:1}' | zq -z 'b*' -
```
produces
```mdtest-output
{s:"bar"}
{s:"baz"}
```
Likewise, a suffix match may be performed as follows:
```mdtest-command
echo '"foo" {s:"bar"} {s:"baz"} {foo:1}' | zq -z '*z' -
```
produces
```mdtest-output
{s:"baz"}
```
and
```mdtest-command
echo '"foo" {s:"bar"} {s:"baz"} {a:1}' | zq -z '*a*' -
```
produces
```mdtest-output
{s:"bar"}
{s:"baz"}
{a:1}
```

Note that a glob may look like multiplication but context disambiguates
these condition, e.g.,
```
a*b
```
is a glob match for any matching string value in the input, but
```
a*b==c
```
is a Boolean comparison between the product `a*b` and `c`.

### 7.2 Search Logic

The search patterns described above can be combined with other elements

> to a search expression comprised of "search terms" that may be combined

Reviewer 2: /to a/to form a/, maybe?

using Boolean logic.

> Note that when processing ZNG data, the Zed runtime performs a multi-threaded
> Boyer-Moore scan over decompressed data buffers before parsing any data.
> This allows large buffers of data to be efficiently discarded and skipped when
> searching for rarely occurring values.  For a [Zed lake](../zed), search indexes
> may also be configured to further accelerate searches.
> In a forthcoming release, Zed will also offer an approach for locating
> delimited words within string fields, which will allow accelerated
> search using a full-text search index.  Currently, search indexes may be built

> > for exact value match as text segmentation is in the works.a

Reviewer 2: ```suggestion
> for exact value match as text segmentation is in the works.
```

that currently perform shaping are:

* [cast](functions/cast.md) - coerce a value to a different type
* [crop](functions/crop.md) - remove fields from a value that are missing in a specified type
* [fill](functions/fill.md) - add null values for missing fields
* [order](functions/order.md) - reorder record fields
* [shape](functions/shape.md) - apply cast, fill, and order

They all have the same signature, taking two parameters: the value to be
transformed and a type value for the target type.

> Another type of transformation that's needed for shaping is renaming fields,
> which is supported by the [rename operator](operators/rename.md#operator).
> Also, the [yield operator](operators/yield.md#operator)
> is handy for simply emitting new, arbitrary record literals based on
> input values and mixing in these shaping functions in an embedded record literal.
> The [fuse aggregate function](aggregates/fuse.md) is useful for fusing
> values into a common schema within a group-by schema though a type is returned
> rather than values.

In the examples below, we will use the following named type `connection`
that is stored in a file `connection.zed`

> and is included in the example Zed queries with `-I` option of `zq`:

Reviewer 2: ```suggestion
and is included in the example Zed queries with the `-I` option of `zq`:
```
